Replace Gson library with Moshi

cbeyls opened the pull request:
**! ! Warning**: Do not merge before testing every API call and database read involving JSON !

**Gson** is obsolete and has been superseded by **Moshi**. But more importantly, parsing Kotlin objects using Gson is _dangerous_ because Gson uses Java serialization and is **not Kotlin-aware**. This has two main consequences:

- Fields of non-null types may end up null at runtime. Parsing will succeed, but the code may crash later with a `NullPointerException` when trying to access a field member;
- Default values of constructor parameters are always ignored. When absent, reference types will be null, booleans will be false and integers will be zero.

On the other hand, Kotlin-aware parsers like **Moshi** or **Kotlin Serialization** will validate at parsing time that all received fields comply with the Kotlin contract and avoid errors at runtime, making apps more stable and schema mismatches easier to detect (as long as logs are accessible):

- Receiving a null value for a non-null type will generate a parsing error;
- Optional types are declared explicitly by adding a default value. **A missing value with no default value declaration will generate a parsing error.**

Migrating the entity declarations from Gson to Moshi will make the code more robust but is not an easy task because of the semantic differences.

With Gson, both nullable and optional fields are represented with a null value. After converting to Moshi, some nullable entities can become non-null with a default value (if they are optional and not nullable), others can stay nullable with no default value (if they are mandatory and nullable), and others can become **nullable with a default value of null** (if they are optional _or_ nullable _or_ both). That third option is the safest bet when it's not clear if a field is optional or not, except for lists which can usually be declared as non-null with a default value of an empty list (I have yet to see a nullable array type in the Mastodon API).

Fields that are currently declared as non-null present another challenge. In theory, they should remain as-is and everything will work fine. In practice, **because Gson is not aware of nullable types at all**, it's possible that some non-null fields currently hold a null value in some cases but the app does not report any error because the field is not accessed by Kotlin code in that scenario. After migrating to Moshi however, parsing such a field will now fail early if a null value or no value is received.

These fields will have to be identified by heavily testing the app and looking for parsing errors (`JsonDataException`) and/or by going through the Mastodon documentation. A default value needs to be added for missing optional fields, and their type could optionally be changed to nullable, depending on the case.

Gson is also currently used to serialize and deserialize objects to and from the local database, which is also challenging because backwards compatibility needs to be preserved. Fortunately, by default Gson omits writing null fields, so a field of type `List<T>?` could be replaced with a field of type `List<T>` with a default value of `emptyList()` and reading back the old data should still work. However, nullable lists that are written directly (not as a field of another object) will still be serialized to JSON as `"null"` so the deserializing code must still be handling null properly.

Finally, changing the database schema is out of scope for this pull request, so database entities that also happen to be serialized with Gson will keep their original types even if they could be made non-null as an improvement.

In the end this is all for the best, because the app will be more reliable and errors will be easier to detect by showing up earlier with a clear error message. Not to mention the performance benefits of using Moshi compared to Gson.

- Replace Gson reflection with Moshi Kotlin codegen to generate all parsers at compile time.
- Replace custom `Rfc3339DateJsonAdapter` with the one provided by moshi-adapters.
- Replace custom `JsonDeserializer` classes for Enum types with `EnumJsonAdapter.create(T).withUnknownFallback()` from moshi-adapters to support fallback values.
- Replace `GuardedBooleanAdapter` with the more generic `GuardedAdapter` which works with any type. Any nullable field may now be annotated with `@Guarded`.
- Remove Proguard rules related to Json entities. Each Json entity needs to be annotated with `@JsonClass` with no exception, and adding this annotation will ensure that R8/Proguard will handle the entities properly.
- Replace some nullable Boolean fields with non-null Boolean fields with a default value where possible.
- Replace some nullable list fields with non-null list fields with a default value of `emptyList()` where possible.
- Update `TimelineDao` to perform all Json conversions internally using `Converters` so no Gson or Moshi instance has to be passed to its methods.
- ~~Create a custom `DraftAttachmentJsonAdapter` to serialize and deserialize `DraftAttachment` which is a special entity that supports more than one json name per field. A custom adapter is necessary because there is not direct equivalent of `@SerializedName(alternate = [...])` in Moshi.~~ Remove alternate names for some `DraftAttachment` fields which were used as a workaround to deserialize local data in 2-years old builds of Tusky.
- Update tests to make them work with Moshi.
- Simplify a few `equals()` implementations.
- Change a few functions to `val`s
- Turn `NetworkModule` into an `object` (since it contains no abstract methods).

Please test the app thoroughly before merging. There may be some fields currently declared as mandatory that are actually optional.

charlag commented:
If we are going to use codegen anyway, why not Kotlin Serialization?

charlag commented:
this is genuine question btw, amazing work too!

cbeyls commented:
> If we are going to use codegen anyway, why not Kotlin Serialization?

I don't know kotlinx serialization well enough to be able to write a custom streaming deserializer or implement something equivalent to `GuardedAdapter`. But If someone wants to do it, be my guest :smile:

Goooler commented:
It could be something like
```kt
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializer
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

@Serializer(forClass = T::class)
object GuardedSerializer : KSerializer<T?> {
    private val json = Json { ignoreUnknownKeys = true }

    override val descriptor = SerialDescriptor("WithCustomDefault", StructureKind.OBJECT)

    override fun serialize(encoder: Encoder, value: T?) {
        // Write the value to the encoder
        encoder.encodeSerializableValue(json.encodeToJsonElement(value))
    }

    override fun deserialize(decoder: Decoder): T? {
        return try {
            // Try to decode the value
            json.decodeFromJsonElement(decoder.decodeSerializableValue(JsonElement.serializer()))
        } catch (e: Exception) {
            // If there's an error, return null
            null
        }
    }
}
```

connyduck reviewed (changes requested):
I found one bug with my mastodon.social account

`Failed to load timeline                                                        com.squareup.moshi.JsonDataException: Required value 'keywords' missing at $[0].filtered[0].filter
`

cbeyls commented on app/src/main/java/com/keylesspalace/tusky/db/DraftEntity.kt:
```diff
@@ -46,24 +52,109 @@ data class DraftEntity(
+ * Tusky 15: uriString = e, description = f, type = g
+ * Tusky 16 beta: uriString = i, description = j, type = k
+ */
+class DraftAttachmentJsonAdapter(moshi: Moshi) : JsonAdapter<DraftAttachment>() {
```
This is indeed legacy code that adds unnecessary complexity to the project. I'll remove it.

cbeyls commented:
> I found one bug with my mastodon.social account
> 
> `Failed to load timeline com.squareup.moshi.JsonDataException: Required value 'keywords' missing at $[0].filtered[0].filter `

I made the type optional with a default value of empty list. This is one of the cases where a non-null Kotlin field ends up null at runtime with Gson. I already fixed one or two more in the initial code I submitted.

cbeyls commented:
> If we are going to use codegen anyway, why not Kotlin Serialization?

Note that contrary to Moshi, the Kotlin Serialization adapter for Retrofit currently does not support streaming: the entire network response is first loaded into a big String, then that String is parsed. Moshi works directly at the byte stream level using Okio, which makes it more efficient in theory.

charlag commented:
this is a good point, thank you

connyduck commented:
> > I found one bug with my mastodon.social account
> > `Failed to load timeline com.squareup.moshi.JsonDataException: Required value 'keywords' missing at $[0].filtered[0].filter `
> 
> I made the "keywords" field optional with a default value of empty list. This is one of the cases where a non-null Kotlin field ends up null at runtime with Gson. I already fixed one or two more in the initial code I submitted.

Now I'm getting `com.squareup.moshi.JsonDataException: Non-null value 'statusMatches' (JSON name 'status_matches') was null at $[1].filtered[0].status_matches`

cbeyls commented:
> Now I'm getting `com.squareup.moshi.JsonDataException: Non-null value 'statusMatches' (JSON name 'status_matches') was null at $[1].filtered[0].status_matches`

The two list fields in `FilterResult` (currently unused in the code) are actually nullable according to the [API documentation](https://docs.joinmastodon.org/entities/FilterResult/). I changed their declaration to reflect that. I need to check if other nullable lists exist in the Mastodon API.

connyduck commented:
Loading the timeline works now and I can't find any other problems!

If they are unused we should probably comment them out to save some deserializing time and memory.

cbeyls commented:
I went trough the entire Mastodon API entities documentation to find remaining optional fields that were declared as mandatory, as well as nullable fields that were declared as non-null.
I'm now quite confident that the code is stable and ready to be merged after a last round of tests.

Tak commented:
Huh, I'm still getting the `Filter.keywords` exception after merging 6be3f0730746d05d77356b90293f0f67c864ff97 locally - if I fix that, everything appears to be working normally for me

cbeyls commented:
> Huh, I'm still getting the `Filter.keywords` exception after merging [6be3f07](https://github.com/tuskyapp/Tusky/commit/6be3f0730746d05d77356b90293f0f67c864ff97) locally - if I fix that, everything appears to be working normally for me

Oops, I just realized I reverted my own change in the recent commit. Turns out the documentation says this field is mandatory but it's in fact optional. I added a comment to avoid a future mistake.
